src/Network.cpp

```diff
@@ -1000,81 +991,88 @@ void Network::show_heatmap(const FastState* const state,
         }
     }
 }
 
-void Network::gather_features(const GameState* const state, NNPlanes & planes) {
-    planes.resize(INPUT_CHANNELS);
-    auto& black_to_move = planes[2 * INPUT_MOVES];
-    auto& white_to_move = planes[2 * INPUT_MOVES + 1];
+void Network::get_input_moves(const GameState* const state, 
+                              std::vector<net_t>& input_data, 
+                              const int rotation, 
+                              const int color)
+{
+    assert(rotation >= 0 && rotation <= 7);
+    const auto moves = std::min<size_t>(state->get_movenum() + 1, INPUT_MOVES);
 
+    for (auto c = size_t{0}; c < moves; c++) {
+        auto const board = state->get_past_board(c);
+        for (auto idx = 0; idx < BOARD_SQUARES; idx++) {
+            const auto rot_vtx = rotate_nn_idx_vtx_table[rotation][idx];
+            const auto color_is_at_vtx = (color == board.get_square(rot_vtx));
+            input_data.emplace_back(net_t(color_is_at_vtx));
+        }
+    }
+    for (auto c = moves; c < INPUT_MOVES; c++) {
+        input_data.insert(input_data.cend(), BOARD_SQUARES, net_t(false));
+    }
+}
+
+void Network::gather_features_vector(const GameState* const state,
+                                     std::vector<net_t>& input_data, 
+                                     const int rotation) {
     const auto to_move = state->get_to_move();
     const auto blacks_move = to_move == FastBoard::BLACK;
+    const auto not_to_move = blacks_move ? FastBoard::WHITE : FastBoard::BLACK;
 
-    const auto black_offset = blacks_move ? 0 : INPUT_MOVES;
-    const auto white_offset = blacks_move ? INPUT_MOVES : 0;
+    input_data.reserve(INPUT_CHANNELS * BOARD_SQUARES);
 
-    if (blacks_move) {
-        black_to_move.set();
-    } else {
-        white_to_move.set();
-    }
+    get_input_moves(state, input_data, rotation, to_move);
+    get_input_moves(state, input_data, rotation, not_to_move);
 
-    const auto moves = std::min<size_t>(state->get_movenum() + 1, INPUT_MOVES);
-    // Go back in time, fill history boards
-    for (auto h = size_t{0}; h < moves; h++) {
-        // collect white, black occupation planes
-        fill_input_plane_pair(state->get_past_board(h),
-                              planes[black_offset + h],
-                              planes[white_offset + h]);
-    }
+    input_data.insert(input_data.cend(), BOARD_SQUARES, net_t(blacks_move));
+    input_data.insert(input_data.cend(), BOARD_SQUARES, net_t(!blacks_move));
 }
 
-int Network::rotate_nn_idx(const int vertex, int symmetry) {
-    assert(vertex >= 0 && vertex < BOARD_SQUARES);
-    assert(symmetry >= 0 && symmetry < 8);
-    auto x = vertex % BOARD_SIZE;
-    auto y = vertex / BOARD_SIZE;
-    int newx;
-    int newy;
-
-    if (symmetry >= 4) {
-        std::swap(x, y);
-        symmetry -= 4;
+int Network::get_board_vertex(const GameState& state, 
+                              const int x, const int y, int rotation) {
```

Comment:
I much prefer the vertex rotation (err, symmetry) and "obtain board contents" be kept split. This has been a source of subtle bugs early on and it could use unit tests, which merging is going to make more problematic.

Reply:
Understood, thanks for the warning about potential subtle bugs. Are you ok with the "obtaining of board vertex" to still be included within the symmetry table initialisation?